File: Oryx.Cognite/src/Beta/DataPoints.fs
// Copyright 2026 Cognite AS
// SPDX-License-Identifier: Apache-2.0

namespace Oryx.Cognite.Beta

open System.IO.Compression

open Com.Cognite.V1.Timeseries.Proto
open Oryx
open Oryx.Cognite
open Oryx.Cognite.Beta

open CogniteSdk

/// Time series data point handlers that opt in to the beta API
[<RequireQualifiedAccess>]
module DataPoints =
    [<Literal>]
    let Url = "/timeseries/data"

    /// Retrieves data points using the beta API
    let list (query: DataPointsQuery) (source: HttpHandler<unit>) : HttpHandler<DataPointListResponse> =
        source
        |> withLogMessage "Beta:DataPoints:list"
        |> withBetaHeader
        |> listProtobuf query Url DataPointListResponse.Parser.ParseFrom

    /// Retrieves aggregate data points using the beta API
    let listAggregates (query: DataPointsQuery) (source: HttpHandler<unit>) : HttpHandler<DataPointListResponse> =
        source
        |> withLogMessage "Beta:DataPoints:listAggregates"
        |> withBetaHeader
        |> listProtobuf query Url DataPointListResponse.Parser.ParseFrom

    /// Insert data points using the beta API
    let create (items: DataPointInsertionRequest) (source: HttpHandler<unit>) : HttpHandler<EmptyResponse> =
        source
        |> withLogMessage "Beta:DataPoints:create"
        |> withBetaHeader
        |> createProtobuf items Url

    /// Insert data points with gzip compression using the beta API
    let createWithGzip
        (items: DataPointInsertionRequest)
        (compression: CompressionLevel)
        (source: HttpHandler<unit>)
        : HttpHandler<EmptyResponse> =
        source
        |> withLogMessage "Beta:DataPoints:create"
        |> withBetaHeader
        |> createGzipProtobuf items compression Url

File: CogniteSdk/src/Resources/Beta/DataPoints.cs
// Copyright 2026 Cognite AS
// SPDX-License-Identifier: Apache-2.0

using System;
using System.IO.Compression;
using System.Threading;
using System.Threading.Tasks;
using Com.Cognite.V1.Timeseries.Proto;
using Microsoft.FSharp.Core;
using Oryx;

namespace CogniteSdk.Resources.Beta
{
    /// <summary>
    /// Beta time series data point operations
    /// </summary>
    public class DataPointsResource : Resource
    {
        /// <summary>
        /// Will only be instantiated by the client
        /// </summary>
        /// <param name="authHandler">Authentication handler.</param>
        /// <param name="ctx">The HTTP context to use for the request.</param>
        internal DataPointsResource(Func<CancellationToken, Task<string>> authHandler, FSharpFunc<IHttpNext<Unit>, Task<Unit>> ctx) : base(authHandler, ctx)
        {
        }

        /// <summary>
        /// Retrieve data points using the beta API
        /// </summary>
        /// <param name="query">Data points query.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<DataPointListResponse> ListAsync(DataPointsQuery query, CancellationToken token = default)
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = Oryx.Cognite.Beta.DataPoints.list(query, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieve the latest data points using the beta API
        /// </summary>
        /// <param name="query">The latest query.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<DataPointListResponse> LatestAsync(DataPointsLatestQuery query, CancellationToken token = default)
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = Oryx.Cognite.Beta.DataPoints.latest(query, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Insert data points using the beta API
        /// </summary>
        /// <param name="points">Data points to insert.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> CreateAsync(DataPointInsertionRequest points, CancellationToken token = default)
        {
            if (points is null)
            {
                throw new ArgumentNullException(nameof(points));
            }

            var req = Oryx.Cognite.Beta.DataPoints.create(points, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }

        /// <summary>
        /// Insert data points using the beta API, gzip-compressed at the given level
        /// </summary>
        /// <param name="points">Data points to insert.</param>
        /// <param name="compression">Gzip compression level.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> CreateAsync(
            DataPointInsertionRequest points,
            CompressionLevel compression,
            CancellationToken token = default)
        {
            if (points is null)
            {
                throw new ArgumentNullException(nameof(points));
            }

            var req = Oryx.Cognite.Beta.DataPoints.createWithGzip(points, compression, GetContext(token));
            return await RunAsync(req).ConfigureAwait(false);
        }
    }
}
